core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/SamplesReceivedWorkflow.java

```diff
@@ -0,0 +1,536 @@
+    }
+
+    @Override
+    public void processInput(PositiveDoubleProgressStep step) {
+      volumeStep = step;
+    }
+
+    public double getVolume() {
+      return volumeStep.getInput();
+    }
+
+  }
+
+  private static class BoxPositionStep implements WorkflowStep {
+
+    private BoxPositionProgressStep boxPositionStep;
+
+    private final int aliquotIndex;
+    private final Box box;
+    private String firstFreePosition;
+    private final Set<String> occupiedLocations;
+
+    private String error;
+
+    BoxPositionStep(int aliquotIndex, Box box, Set<String> occupiedLocations) {
+      this.aliquotIndex = aliquotIndex;
+      this.box = box;
+      this.occupiedLocations = occupiedLocations;
+    }
+
+    @Override
+    public WorkflowStepPrompt getPrompt() {
+      firstFreePosition = getFirstFreePosition();
+      return new WorkflowStepPrompt(Sets.newHashSet(InputType.BOX_POSITION, InputType.SKIP),
+          String.format("Enter a Box Position to place Aliquot #%d.\nSkipping will default to position %s", aliquotIndex + 1,
+              firstFreePosition));
+    }
+
+    @Override
+    public ProgressStep getProgressStep() {
+      return boxPositionStep;
+    }
+
+    @Override
+    public void cancelInput() {
+      occupiedLocations.remove(boxPositionStep.getInput());
+      boxPositionStep = null;
+    }
+
+    @Override
+    public String getLogMessage() {
+      return String.format("Place Aliquot #%d at position %s", aliquotIndex + 1, getPosition());
+    }
+
+    @Override
+    public void processInput(BoxPositionProgressStep step) {
+      if (isLocationFree(step.getInput())) {
+        boxPositionStep = step;
+        occupiedLocations.add(getPosition());
+      } else if (!box.isValidPosition(step.getInput()) || !isColumnNonZero(step.getInput())) {
+        error = String.format("The Box '%s' does not have a position '%s'", box.getAlias(), step.getInput());
+      } else if (!box.isFreePosition(step.getInput()) || occupiedLocations.contains(step.getInput())) {
+        error = String.format("The position '%s' is already occupied", step.getInput());
+      }
+    }
+
+    @Override
+    public void processInput(SkipProgressStep step) {
+      this.boxPositionStep = new BoxPositionProgressStep();
+      boxPositionStep.setInput(firstFreePosition);
+      occupiedLocations.add(getPosition());
+    }
+
+    public String getPosition() {
+      return boxPositionStep != null ? boxPositionStep.getInput() : firstFreePosition;
+    }
+
+    public boolean isLocationFree(String position) {
+      return (boxPositionStep != null && boxPositionStep.getInput().equals(position)) ||
+          (box.isValidPosition(position) && isColumnNonZero(position) && box.isFreePosition(position)
+              && !occupiedLocations.contains(position));
+      // box.isFreePosition does not seem to work properly when the workflow is viewed through the view workflows tab
```

Comment:
I'm unclear as to the purpose of this comment -- is this a TODO for you or a warning for others who may be reading the code later?

Reply:
It's the strange SQL error that me and @apmasell need to look into, so a TODO for me. It will be removed once it is fixed.